lib/dRep/dRepImpl.py

```diff
@@ -0,0 +1,441 @@
+
+    #BEGIN_CLASS_HEADER
+    #END_CLASS_HEADER
+
+    # config contains contents of config file in a hash or None if it couldn't
+    # be found
+    def __init__(self, config):
+        #BEGIN_CONSTRUCTOR
+
+        callback_url = os.environ['SDK_CALLBACK_URL']
+        workspace_url = config['workspace-url']
+        shared_folder = config['scratch']
+        
+        self._globals = { # shared by all API-method runs
+            'shared_folder': config['scratch'], 
+            'ws': Workspace(workspace_url),
+            'dfu': DataFileUtil(callback_url),
+            'mgu': MetagenomeUtils(callback_url),
+            'kbr': KBaseReport(callback_url),
+            }
+
+        logging.basicConfig(format='%(created)s %(levelname)s: %(message)s',
+                            level=logging.INFO)
+        #END_CONSTRUCTOR
+        pass
+
+
+    def dereplicate(self, ctx, params):
+        """
+        This example function accepts any number of parameters and returns results in a KBaseReport
+        :param params: instance of mapping from String to unspecified object
+        :returns: instance of type "ReportResults" -> structure: parameter
+           "report_name" of String, parameter "report_ref" of String
+        """
+        # ctx is the context object
+        # return variables are: output
+ 
+        #BEGIN dereplicate
+
+        dprint('params', run=locals())
+
+
+        # set up globals ds `_globals` for this API-method run
+
+        reset(_globals) # clear all fields but `debug`
+        _globals.update({
+            **self._globals,
+            'workspace_name': params['workspace_name'],
+            'run_dir': os.path.join(self._globals['shared_folder'], str(uuid.uuid4())),
+            'warnings': [],
+            })
+
+        os.mkdir(_globals.run_dir)
+
+
+
+
+        #
+        ##
+        ###
+        ####
+        #####
+
+
+        def simple_return(msg, report_params_kwargs={}):
+            '''
+            Return with a simple message
+            '''
+            report_params = {
+                    'message': msg,
+                    'workspace_name': _globals.workspace_name,
+                    **report_params_kwargs
+                    }
+            report_info = _globals.kbr.create_extended_report(report_params)
+            output = {
+                'report_name': report_info['name'],
+                'report_ref': report_info['ref'],
+            }
+
+            return [output]
+
+
+
+        # 
+        ##
+        ### check unique UPAs
+        #### 
+        #####
+
+        if len(set(params['genomes_refs'])) < len(params['genomes_refs']):
+            msg = 'Removing duplicate input BinnedContigs'
+            logging.warning(msg)
+            _globals.warnings.append(msg) 
+            params['genomes_refs'] = list(dict.fromkeys(params['genomes_refs']))
+
+       
+
+
+
+        # 
+        ##
+        ### load BinnedContigs files (to scratch) and obj data
```

Comment:
The comments with description text are definitely helpful. The triangle shapes don't really help the reader, but they do give me finger exercise scrolling